Feature: Group integration

Adnan-cds opened the pull request:
Provides:
- A Group plugin for Alert banners.
- A View for listing Alert banners belonging to Groups.
- User permissions to go along with the localgov_microsites_group module.

Adnan-cds committed:
Feat: Group integration.

Provides:
- A Group plugin for Alert banners.
- A View for listing Alert banners belonging to groups.
- User permissions to go along with the localgov_microsites_group module.

Adnan-cds committed:
Fix: Group integration.

- Fixed Alert banner's entity *type* name during config dependency calculation.
- Added group relation type cached clear during Alert banner bundle creation.
  This is needed for config import to work.

Adnan-cds committed:
Test: Group integration.

Automated test for banner allocation to a test group bundle.

Adnan-cds committed:
Fix: Coding standards.

Adnan-cds committed:
Test: Requiring suggested dependencies.

The suggested dependencies (e.g. group) from composer.json are required to run
some of the tests.

Adnan-cds committed:
Test: Requiring suggested dependencies.

Fix for composer command in test runner to grab suggested packages.

Adnan-cds commented:
Mostly [lifted](https://git.drupalcode.org/project/group/-/blob/3.0.x/modules/gnode/src/Plugin/Group/Relation/GroupNode.php) from the gnode submodule of the group module.

andybroomfield commented on modules/group_alert_banner/group_alert_banner.group.permissions.yml:
```diff
@@ -0,0 +1,3 @@
+access localgov_alert_banner overview:
+  title: 'Access Alert banner listing page'
+  description: 'Access the Alert banner listing and management page, regardless of banner type.'
```
This sounds like the kind of general permission we should have in the main module, how are we access controlling it at the moment? Why do we need the seperate permission?

andybroomfield replied:
So I think I'm understanding that this is for the seperate group view, which is different from the main admin view. And we only want to grand group insiders the group admin view and not the main one?
(This is me trying to understand localgov microsites permissions)

Adnan-cds replied:
> And we only want to grant group insiders the group admin view and not the main one?

That's right.

andybroomfield commented:
As this is for microsites I'm happy to leave the review to those using it.
Be good to be able to have banners per group.
Raises a couple of questions on permissions, see comment above, though that might just be I don't have experince with group permissions.

andybroomfield commented on modules/group_alert_banner/README.md:
```diff
@@ -0,0 +1,12 @@
+If you are using this module alongside the [localgov_microsites_group](https://github.com/localgovdrupal/localgov_microsites_group) module, then user permissions will be automatically setup during module installation.
+
+Otherwise enable at least the following permissions for **Groups**:
+- "Entity: View any alert banner entities" for both anonymous and authenticated users.
```
May have an issue with #219 and #256 if we remove this permission.

andybroomfield commented on modules/group_alert_banner/group_alert_banner.module:
```diff
@@ -0,0 +1,82 @@
+        'view all localgov alert banner entities',
+        'view all localgov alert banner entity pages',
+      ],
+      Drupal\localgov_microsites_group\RolesHelper::MICROSITES_EDITOR_ROLE => [
+        'access localgov alert banner listing page',
+        'use localgov_alert_banners transition create_new_draft',
+        'use localgov_alert_banners transition publish',
+        'use localgov_alert_banners transition unpublish',
+        'view all localgov alert banner entities',
+        'view all localgov alert banner entity pages',
+      ],
+    ],
```
Granting this to the editors would be a major change in the permisions model, which is that alerts are by the emergency publisher. 
(Although I'm aware of councils that change this to also allow editors).
Is the emergency publisher role not something avalible on microsites?

andybroomfield commented:
My very quick run through within the localgov microsites distrubtion shows that this works well, thanks @Adnan-cds.
I think there are a few oddities, which could be a combination of microsites oddities, issues with domains, and understanding the group permissions vs global permissions. 

Permissions are something we need to look at, just to make sure we're not granting a bit more than we should. 
And we should redirect users after putting an alert banner live to the group version of the alert banner admin view.

I'm also noticing that when I click set live when creating a new alert banner, I'm taken to the view alert page, not the confirm form.

stephen-cox reviewed (changes requested):
This largely works fine. A couple of minor issues:

- As Andy points out, when adding alert the set live option doesn't work.
- When viewing the list of alert banners to the banner type is empty; maybe this isn't needed.

Just to note, that this branch is rather out of data and alert banners weren't displaying until I merged the latest changes in the 1.x branch.

If the add alert banner form redirects to the site live form when the set live field is selected then I think this is fine to merge.

Adnan-cds committed:
Fix: Group banner publishing.

When a new banner is set to go live, we should end up in the "Put banner live"
tab of the new banner after the banner form is submitted.  At the moment we are
ending up in the banner's "View" tab.

Also, group banners should only appear in group pages and not all over the site.

Adnan-cds committed:
Fix: Account for Domain group.

Decide the active group either based on the current path or the current domain.
This is useful where we are using the domain_group module to associate a domain
with a group.

Adnan-cds committed:
Merge remote-tracking branch 'origin/1.x' into feature/group-integration

Adnan-cds committed:
Chore: Coding standards fix.

Adnan-cds committed:
Fix: Post-banner publishing redirection.

Once a group banner is published, redirect to the group banner listing page
instead of to the site-wide banner listing page.

Adnan-cds commented:
> And we should redirect users after putting an alert banner live to the group version of the alert banner admin view.

This is done now.

> when I click set live when creating a new alert banner, I'm taken to the view alert page, not the confirm form.

This is fixed as well.

The test runs are resulting in an [unrelated test failure](https://github.com/localgovdrupal/localgov_alert_banner/actions/runs/6074964785/job/16480308511?pr=250#step:5:21).  Is this a known issue?

Adnan-cds committed:
Merge remote-tracking branch 'origin/1.x' into feature/group-integration

finnlewis commented:
@Adnan-cds would like @andybroomfield to review before merging. 

@andybroomfield would like @stephen-cox to review before merging! 😄

finnlewis commented:
@stephen-cox -  might you have time to review this one? 
@ekes  -  might you have time to review this one? 

@finnlewis can you test again?

and @millnut  - if you get bored?  😄

stephen-cox reviewed (approved):
Working well for me. Happy to approve.

One things to note is to get alert banners to display I had to add the alert banner block to the themes header region. I suggest adding optional block config for this to the Microsites Base theme as we have for other blocks: https://github.com/localgovdrupal/localgov_microsites_base/tree/2.x/config/install

Adnan-cds commented:
>  I suggest adding optional block config for this to the Microsites Base theme as we have for other blocks

That's a good point that I missed.  I remember having to manually add this block in our microsite instance.  How about adding it here in the group_alert_banner submodule as an optional block config?

stephen-cox commented:
> How about adding it here in the group_alert_banner submodule as an optional block config?

That would do the trick

Adnan-cds committed:
Merge remote-tracking branch 'origin/1.x' into feature/group-integration

Adnan-cds committed:
Feat: Group banner block.

Optional Alert banner block configuration for the localgov_microsites_base
theme.  Automates setup of the Alert banner block for LocalGov Drupal Microsites
distribution users.

Adnan-cds commented:
I have added an [optional config file](https://github.com/localgovdrupal/localgov_alert_banner/pull/250/files#diff-6303ce6fd5d4ee47488998ea4ab3a415a1e0058ea7afc8139b7750a4650ff11c) to add the Alert banner block in the localgov_microsites_base theme.  I have also explained the [block placement issue in the README](https://github.com/localgovdrupal/localgov_alert_banner/blob/feature/group-integration/modules/group_alert_banner/README.md#block-placement).  Please let me know if this will do.

stephen-cox commented:
Looks fine to me :+1:

Adnan-cds commented:
Great :) Merging then...
